RHDEVDOCS-3678 - Add section for direct access to 3rd party monitoring components' service APIs (#41384)
openshift/openshift-docs: modules/monitoring-accessing-third-party-monitoring-web-service-apis.adoc

// Module included in the following assemblies:
//
// * monitoring/accessing-third-party-monitoring-uis-and-apis.adoc

:_content-type: PROCEDURE
[id="accessing-third-party-monitoring-web-service-apis"]
= Accessing third-party monitoring web service APIs

[role="_abstract"]

> You can directly access third-party web service APIs from the command line for monitoring stack components such as Prometheus, Alertmanager, Thanos Ruler, and Thanos Querier. 

Reviewer: no need to mention Prometheus here, since Prometheus UI is still could be accessed in 4.10

Reviewer 2: > no need to mention Prometheus here, since Prometheus UI is still could be accessed in 4.10

Prometheus API is also accessible.

Author: I'm keeping the reference to Prometheus because it's API is also accessible.

This access only supports using a Bearer Token for authentication. 

> 

Author: @raptorsun Can you suggest some wording here for adding the information you gave me about the minimum permissions needed in Oauth-proxy for Alertmanager?

Reviewer 3: The access setting comes from [this argument](https://github.com/openshift/cluster-monitoring-operator/blob/master/assets/alertmanager/alertmanager.yaml#L32). We may explain the setting as below:

`Openshift uses Subject Access Review for Alertmanager. It requires the user to have at least "patch" permission to resource "alertmanager" in namespace "openshift-monitoring".`

Could you please double check? @jan--f  I'm not so familiar with [Subject Access Review](https://kubernetes.io/docs/reference/access-authn-authz/authorization/) and may miss some details 😅

Reviewer 4: It might be more convenient to say that the associated account needs to be bound against the `monitoring-alertmanager-edit` role in the `openshift-monitoring` namespace (this way the exact permissions may change in the future without having to reword the documentation).

Reviewer 3: Agree. We just tell users to bond to this role instead of listing permissions. Easier to understand for users.

Author: @simonpasquier @raptorsun Is this same account binding also required for the components other than Alertmanager? Or do these other components require that the account be bound to a different role?

Reviewer: 1. I am afraid the `monitoring-alertmanager-edit` role is not enough to see the alertmanager APIs, the role is not able to execute `host=$(oc -n openshift-monitoring get route alertmanager-main -ojsonpath={.spec.host})`, if the user know the alertmanager route, then it is not needed to get the route, there is not such issue
2. The `monitoring-alertmanager-edit` role is not able to view thanos-querier service APIs
```
# oc adm policy add-role-to-user monitoring-alertmanager-edit testuser-11 -n openshift-monitoring --role-namespace openshift-monitoring
role.rbac.authorization.k8s.io/monitoring-alertmanager-edit added: "testuser-11"

# oc login https://api.***.com:6443 -u testuser-11 -p ***
The server uses a certificate signed by an unknown authority.
You can bypass the certificate check, but any data you send to the server could be intercepted by others.
Use insecure connections? (y/n): y

Login successful.

You don't have any projects. You can try to create a new project, by running

    oc new-project <projectname>

# host=$(oc -n openshift-monitoring get route alertmanager-main -ojsonpath={.spec.host})
Error from server (Forbidden): routes.route.openshift.io "alertmanager-main" is forbidden: User "testuser-11" cannot get resource "routes" in API group "route.openshift.io" in the namespace "openshift-monitoring"

### if we already know the alertmanager route, no such issue:
# host=alertmanager-main-openshift-monitoring.*****
# token=#(oc whoami -t)
# curl -H "Authorization: Bearer #token" -k "https://#host/api/v2/receivers"
[{"name":"Default"},{"name":"Watchdog"},{"name":"Critical"}]

```

Reviewer 3: - Alertmanager: the role `monitoring-alertmanager-edit` grants access to alert manager API
- Thanos Ruler & Querier: The requesting account must have "get" permission on the namespace that it is querying. The user may create their roles and grant them access to certain namespaces. For us we cannot give an exhaust list of these roles in advance. 
What do you think? @simonpasquier

Reviewer: we should change the sequence to
```
The following example shows how to query the service API receivers for Alertmanager

This access requires that the associated user account be bound against the monitoring-alertmanager-edit role in the openshift-monitoring namespace and only supports using a Bearer Token for authentication.
```
and add note, for other service APIs, the requesting account must have "get" permission on the namespace

Author: I updated the text as follows:

```
The following example shows how to query the service API receivers for Alertmanager.
This example requires that the associated user account be bound against the `monitoring-alertmanager-edit` role in the `openshift-monitoring` namespace and only supports using a Bearer Token for authentication.

[source, terminal]
----
$ host=$(oc -n openshift-monitoring get route alertmanager-main -ojsonpath={.spec.host})
$ token=$(oc whoami -t)
$ curl -H "Authorization: Bearer $token" -k "https://$host/api/v2/receivers"
----

To access Thanos Ruler and Thanos Querier service APIs, the requesting account must have `get` permission for the namespace to be queried.
```

// Module included in the following assemblies:
//
// * monitoring/accessing-third-party-monitoring-uis-and-apis.adoc

:_content-type: PROCEDURE
[id="accessing-third-party-monitoring-web-service-apis"]
= Accessing third-party monitoring web service APIs

[role="_abstract"]
You can directly access third-party web service APIs from the command line for monitoring stack components such as Prometheus, Alertmanager, Thanos Ruler, and Thanos Querier. 

The following example shows how to query the service API receivers for Alertmanager.
This example requires that the associated user account be bound against the `monitoring-alertmanager-edit` role in the `openshift-monitoring` namespace and that the account has the privilege to view the route.
This access only supports using a Bearer Token for authentication.

[source, terminal]
----
$ host=$(oc -n openshift-monitoring get route alertmanager-main -ojsonpath={.spec.host})
$ token=$(oc whoami -t)
$ curl -H "Authorization: Bearer $token" -k "https://$host/api/v2/receivers"
----

> To access Thanos Ruler and Thanos Querier service APIs, the requesting account must have `get` permission for the namespace to be queried.

Reviewer 5: Move this to the prerequisites section?

Author: It's really more of an "additional information" sentence to note a difference with accessing these components versus what is being accessed in the example code. I've set it off with a Note admonition because of how it's functioning here.

// Module included in the following assemblies:
//
// * monitoring/accessing-third-party-monitoring-uis-and-apis.adoc

:_content-type: PROCEDURE
[id="accessing-third-party-monitoring-web-service-apis"]
= Accessing third-party monitoring web service APIs

[role="_abstract"]
You can directly access third-party web service APIs from the command line for monitoring stack components such as Prometheus, Alertmanager, Thanos Ruler, and Thanos Querier. 

The following example shows how to query the service API receivers for Alertmanager.
This example requires that the associated user account be bound against the `monitoring-alertmanager-edit` role in the `openshift-monitoring` namespace and that the account has the privilege to view the route.
This access only supports using a Bearer Token for authentication.

[source, terminal]
----
$ host=$(oc -n openshift-monitoring get route alertmanager-main -ojsonpath={.spec.host})
$ token=$(oc whoami -t)
$ curl -H "Authorization: Bearer $token" -k "https://$host/api/v2/receivers"
----

[NOTE]
====

> To access Thanos Ruler and Thanos Querier service APIs, the requesting account must have `get` permission for the namespace to be queried.

Reviewer 4: We could also mention the `cluster-monitoring-view` clusterrole.
```suggestion
To access Thanos Ruler and Thanos Querier service APIs, the requesting account must have the `get` permission on the namespaces resource.
```

Author: > We could also mention the `cluster-monitoring-view` clusterrole.

@simonpasquier Would this wording work? "To access Thanos Ruler and Thanos Querier service APIs, the requesting account must have the `get` permission on the namespaces resource and be assigned the `cluster-monitoring-view` cluster role."

Reviewer 4: Both parts are synonyms. Maybe: "To access Thanos Ruler and Thanos Querier service APIs, the requesting account must have the `get` permission on the namespaces resource which can be done by granting the `cluster-monitoring-view` cluster role."

Author: Fixed wording per your suggestion
